blog/_posts/2021-08-16-write-more-bash.md

```diff
@@ -0,0 +1,206 @@
+You Can Explore it with
+Curl"](https://blog.tilt.dev/2021/03/18/kubernetes-is-so-simple.html).)
+
+This API platform allows teams to create very complex, robust features.
+But few people appreciate that you can also use it to build simple, hacky features!
+
+A few weeks ago [L](https://twitter.com/ellenkorbes) asked if we could add a
+cancel button to the Tilt UI to kill a running local server. And I realized that you could
+hack this together with a couple dozen lines of Bash without modifying core Tilt! Here it is!
+
+[Cancel](https://github.com/tilt-dev/tilt-extensions/tree/master/cancel) - adds a 'cancel' button for any resource that adds a `local_resource()`.
+
+All you need to do is add this line to your Tiltfile!
+
+```python
+include('ext://cancel')
+```
+
+## What a Bash-based Tilt extension looks like
+
+A reactive control loop looks the same for every feature: 
+
+1. Watch objects
+2. When objects change, check if the expected state matches the desired state.
+3. If they don't match, do stuff
+
+For a cancel button, that means:
+
+1. Watch all local resources
+2. When the local resources change, check if they want cancel buttons or need cancel buttons.
+3. If "want" != "need", do stuff!
+
+OK. To start, we add a new server for adding cancel buttons. We call it `operator:cancel`. Here's
+what [the configuration](https://github.com/tilt-dev/tilt-extensions/blob/master/cancel/Tiltfile) looks like:
+
+```python
+local_resource(
+  name='operator:cancel',
+  serve_cmd='./cancel_btn_controller.sh')
+```
+
+"Operator" and "Controller" are both common words the Kubernetes ecosystem uses for 
+the implementation code behind a feature.
+
+Next, we write [a basic
+loop](https://github.com/tilt-dev/tilt-extensions/blob/master/cancel/cancel_btn_controller.sh)
+in Bash that watches objects for changes:
+
+```bash
+#!/bin/bash
+#
+# Starts a controller that watches for new Cmds, and
+# creates buttons to cancel them.
+
+set -eou pipefail
+
+
+echo "operator:cancel runs in the background and listens to Tilt
+When there are commands to cancel, operator:cancel adds a Cancel button to the Tilt UI
+"
+
+tilt get cmd --watch -o name | while read -r cmd_full_name; do
+    cmd_short_name=${cmd_full_name#cmd.tilt.dev/}
+    ./reconcile_cancel_btn.sh "$cmd_short_name"
+done
+```
+
+This code uses Tilt's [Cmd API](https://api.tilt.dev/core/cmd-v1alpha1.html) to
+watch all local commands. Tilt's CLI reuses large parts of `kubectl`, and has
+many of the same verbs.  You can use `tilt api-resources` to look at all the
+available objects, `tilt get` to view a summary of those objects, and `tilt get
+--watch` to watch for updates to those objects.
+
+This Bash loop streams Cmd changes and passes them to a "reconciler."
+
+"Reconciler" is another common phrase in the Kubernetes ecosystem for the part
+of a controller that compares state to the desired state. You can read more about reconcilers
```

Comment:
1. tie it back to the three steps listed above
2. it also does step (3), yeah?

```suggestion
of a controller that implements steps (2) and (3) of the reactive control loop above: it compares the live state to the desired state, and applies changes to the live state to bring it to the desired state. You can read more about reconcilers
```

Reply:
done